033-OSSDatabaseMigration/Coach/05-online-migration.md

```diff
@@ -0,0 +1,223 @@
+
+This creates a psql dump text file. We need to import it to the target - schema only. Create a separate database for online migration (suggeted ). Alternately you can drop and re-create the database wth.
+
+To drop all the tables with indexes
+
+```bash
+\out drop_tables.sql
+
+select 'drop table ' || tablename || ' cascade;'  from pg_tables where tableowner = 'contosoapp' and schemaname = 'public' ;
+```
+
+To import the schema only to target using psql
+
+```bash
+$ psql -h pgtarget.postgres.database.azure.com -p 5432 -U contosoapp@pgtarget  -d wth <dump_wth.sql
+
+postgres=> create database wth
+```
+
+Import the schema to target
+
+```bash
+psql -h pgtarget.postgres.database.azure.com -U contosoapp@pgtarget -d wth2 < dump_wth.sql
+```
+
+Verify count of tables and indexes created on the target database from psql. 26tables, 39 indices.
+
+```bash
+\dt+
+
+\di+
+
+```
+
+The query to disable all foreign key in DMS  (https://docs.microsoft.com/en-us/azure/dms/tutorial-postgresql-azure-postgresql-online-portal )  shows more results than needed. It can be simplified bu saving the output of a SQL command that just drops the constraints. From psql,
+
+```bash
+\out drop_fk.sql
+```
+
+```sql
+
+SELECT CONCAT('ALTER TABLE ', table_schema, '.', table_name, STRING_AGG(DISTINCT CONCAT(' DROP CONSTRAINT ', foreignkey), ','), ';') as DropQuery
+FROM
+    (SELECT
+    S.table_schema,
+    S.foreignkey,
+    S.table_name,
+    STRING_AGG(DISTINCT S.column_name, ',') AS column_name,
+    S.foreign_table_schema,
+    S.foreign_table_name,
+    STRING_AGG(DISTINCT S.foreign_column_name, ',') AS foreign_column_name
+FROM
+    (SELECT DISTINCT
+    tc.table_schema,
+    tc.constraint_name AS foreignkey,
+    tc.table_name,
+    kcu.column_name,
+    ccu.table_schema AS foreign_table_schema,
+    ccu.table_name AS foreign_table_name,
+    ccu.column_name AS foreign_column_name
+    FROM information_schema.table_constraints AS tc
+    JOIN information_schema.key_column_usage AS kcu ON tc.constraint_name = kcu.constraint_name AND tc.table_schema = kcu.table_schema
+    JOIN information_schema.constraint_column_usage AS ccu ON ccu.constraint_name = tc.constraint_name AND ccu.table_schema = tc.table_schema
+WHERE constraint_type = 'FOREIGN KEY'
+    ) S
+    GROUP BY S.table_schema, S.foreignkey, S.table_name, S.foreign_table_schema, S.foreign_table_name
+    ) Q
+    GROUP BY Q.table_schema, Q.table_name;
+
+```
+
+Edit the file drop_fk.sql to clean up to run as a  SQL statement
+Run the file
+
+```bash
+\i  drop_fk.sql
+```
+
+There is no trigger in the application schema. Still it is best practice to check for it
+
+```sql
+\out drop_trigger.sql
+SELECT DISTINCT CONCAT('ALTER TABLE ', event_object_schema, '.', event_object_table, ' DISABLE TRIGGER ', trigger_name, ';')
+FROM information_schema.triggers
+```
+
+For Azure DMS,  you need to create a migration project
+
+in DMS wizard, connect to source as the external IP from kubectl -n postgresql get svc command
```

Comment:
Capitalize first letter of this sentence.

Reply:
Wrap the `kubectl` command in this paragraph in backticks so it stands out and monospaced.

Reply:
Fixed